Improve join filter pushdown for ON clause conditions

EslamAhmed171 opened the pull request:
## Outline

This PR improves join filter pushdown for conditions in the ON clause in the binder. Previously, the binder categorized join conditions into two groups: join conditions and arbitrary conditions. When the join type was LEFT OUTER, it would push filters that referenced only the right-hand side (RHS), but all other join types missed this optimization opportunity.

## Changes

This PR categorizes join conditions into four categories based on what they reference:

- **Both tables** - References columns from both left and right sides
- **Right only** - References only right-side columns
- **Left only** - References only left-side columns
- **None** - Constants or expressions with no column references

For each category, the binder now makes smarter decisions:

- **NONE**: Check if the expression evaluates to TRUE and can be eliminated. If not, add it to `arbitrary conditions`
- **LEFT**: Check if it can be pushed to the left child. If not, add it to `arbitrary conditions`
- **RIGHT**: Check if it can be pushed to the right child. If not, add it to `arbitrary conditions`
- **BOTH**: Check if it's a comparison that can become a join condition. If not, add it to `arbitrary conditions`

## Correctness Proof

**INNER JOIN**: Both left-side and right-side filters can be pushed down safely. Any tuple that doesn't satisfy the filter won't appear in the result set anyway.

**LEFT OUTER JOIN**: Only right-side filters can be pushed down. Left-side filters would eliminate rows that should appear with NULL-extended right columns.

**RIGHT OUTER JOIN**: Only left-side filters can be pushed down (symmetric to LEFT OUTER).

**SEMI and ANTI JOINs**: These require more careful analysis. Consider a set called `RHS_MATCH_SET` containing all right-side tuples that satisfy the join condition for each left-side tuple.

- *Right-side filters (SEMI and ANTI)*: If a right-side row is filtered out, it won't be in `RHS_MATCH_SET`. This doesn't affect correctness for either SEMI or ANTI joins, so pushing right-side filters is safe.
- *Left-side filters on SEMI JOIN*: If a left tuple doesn't satisfy the filter, `RHS_MATCH_SET` would be empty anyway (the tuple wouldn't match). Early elimination is safe because SEMI only returns left rows with matches.
- *Left-side filters on ANTI JOIN*: If a left tuple doesn't satisfy the filter, `RHS_MATCH_SET` is empty, which means the tuple WOULD be returned by ANTI (no matches = return the row). Pushing the filter would eliminate rows that should be in the result, so it's NOT safe.

## Performance Impact

This optimization allows some joins to use HASH JOIN instead of nested loop join (NLJ), which significantly improves performance depending on relation sizes (10x to 10,000x speedup in some cases).

Using the benchmark from #19213, SEMI JOIN was **2200x slower** than INNER JOIN. After this fix, SEMI JOIN is actually slightly faster than INNER JOIN since it can now use HASH JOIN properly.

## Related Issue

Fixes #19213

EslamAhmed171 committed:
Add filter pushdown helper functions

EslamAhmed171 committed:
Add constant filter evaluation

EslamAhmed171 committed:
Add helper function to check comparison expression

EslamAhmed171 committed:
Refactor join condition extraction with improved filter pushdown

EslamAhmed171 committed:
Improve create join - take more good decision

EslamAhmed171 committed:
Add inner join tests

EslamAhmed171 committed:
Add left outer join tests

EslamAhmed171 committed:
Add ASOF join tests

EslamAhmed171 committed:
Add semi and anti joins tests

EslamAhmed171 committed:
Fix ASOF error message and SEMI join corner case

EslamAhmed171 committed:
Add BOTH-side expression test

EslamAhmed171 committed:
Add right outer join tests

EslamAhmed171 committed:
Add a missing test in ASOF join and minor format

Tmonster reviewed (commented):
Thanks for the PR! Looks pretty good, might want to see a test with an anti join on a condition that always evaluates to true to see what happens.

Is there any reason this could not be done in the optimizer?

lnkuiper commented on src/planner/binder/tableref/plan_joinref.cpp:
```diff
@@ -65,56 +160,36 @@ static bool CreateJoinCondition(Expression &expr, const unordered_set<idx_t> &le
+		auto side = JoinSide::GetJoinSide(*expr, left_bindings, right_bindings);
+
+		if (side == JoinSide::NONE) {
+			if (CanEliminate(context, type, expr)) {
```
Yes, you can do this in a separate PR :)

EslamAhmed171 committed:
Add comments explaining ON clause filter pushdown logic

Clarifies that CanPushToLeftChild and CanPushToRightChild are only
used for ON clause conditions, not WHERE clause filters

EslamAhmed171 committed:
Add ON TRUE condition tests for different join types

Covers INNER, LEFT, RIGHT, FULL, SEMI, and ANTI joins with always-TRUE
conditions. Includes empty table tests for SEMI and ANTI joins.

lnkuiper commented:
Thanks for the changes! LGTM now
